systemtest/src/test/java/io/strimzi/systemtest/migration/MigrationST.java

```diff
@@ -0,0 +1,486 @@
+            .withTopicName(kraftTopicName)
+            .build();
+
+        resourceManager.createResourceWithWait(extensionContext,
+            immediateClients.producerTlsStrimzi(testStorage.getClusterName()),
+            immediateClients.consumerTlsStrimzi(testStorage.getClusterName())
+        );
+
+        ClientUtils.waitForClientsSuccess(testStorage.getProducerName(), testStorage.getConsumerName(), testStorage.getNamespaceName(), testStorage.getMessageCount());
+
+        LOGGER.info("9. migration is completed, waiting for continuous clients to finish");
+        ClientUtils.waitForClientsSuccess(continuousClients.getProducerName(), continuousClients.getConsumerName(), testStorage.getNamespaceName(), continuousClients.getMessageCount());
+    }
+
+    void doRollback(TestStorage testStorage) {
+        LOGGER.info("From {} state we are going to roll back to ZK", KafkaMetadataState.KRaftPostMigration.name());
+
+        LOGGER.info("6. rolling migration back - applying {}: rollback annotation", Annotations.ANNO_STRIMZI_IO_KRAFT);
+        KafkaResource.replaceKafkaResourceInSpecificNamespace(testStorage.getClusterName(),
+            kafka -> kafka.getMetadata().getAnnotations().put(Annotations.ANNO_STRIMZI_IO_KRAFT, "rollback"), testStorage.getNamespaceName());
+
+        LOGGER.info("7. waiting for Broker Pods to be rolled");
+        brokerPodsSnapshot = RollingUpdateUtils.waitTillComponentHasRolledAndPodsReady(testStorage.getNamespaceName(), brokerSelector, 3, brokerPodsSnapshot);
+
+        LOGGER.info("8. waiting until .status.kafkaMetadataState contains {} state", KafkaMetadataState.KRaftDualWriting.name());
+        KafkaUtils.waitUntilKafkaStatusContainsKafkaMetadataState(testStorage.getNamespaceName(), testStorage.getClusterName(), KafkaMetadataState.KRaftDualWriting.name());
+
+        LOGGER.info("9. deleting Controller's NodePool");
+        KafkaNodePool controllerPool = KafkaNodePoolResource.kafkaNodePoolClient().inNamespace(testStorage.getNamespaceName()).withName(testStorage.getControllerPoolName()).get();
+        resourceManager.deleteResource(controllerPool);
+
+        LOGGER.info("10. applying {}: disabled annotation", Annotations.ANNO_STRIMZI_IO_KRAFT);
+        KafkaResource.replaceKafkaResourceInSpecificNamespace(testStorage.getClusterName(),
+            kafka -> kafka.getMetadata().getAnnotations().put(Annotations.ANNO_STRIMZI_IO_KRAFT, "disabled"), testStorage.getNamespaceName());
+
+        LOGGER.info("11. waiting for Broker Pods to be rolled");
+        brokerPodsSnapshot = RollingUpdateUtils.waitTillComponentHasRolledAndPodsReady(testStorage.getNamespaceName(), brokerSelector, 3, brokerPodsSnapshot);
+
+        LOGGER.info("12. waiting until .status.kafkaMetadataState contains {} state", KafkaMetadataState.ZooKeeper.name());
+        KafkaUtils.waitUntilKafkaStatusContainsKafkaMetadataState(testStorage.getNamespaceName(), testStorage.getClusterName(), KafkaMetadataState.ZooKeeper.name());
```

Comment:
one thing we should test ...
During migration, when you reach the `KRaftPostMigration` phase, the brokers become so called "observers" and they get a copy of the `__cluster_metadata` topic from the controllers. It's something you could test that this topic exist during the migration as well.
When you rollback and come back to ZooKeeper, we should test that this topic doesn't exist anymore.
The presence should be tested to check the file system of the Kafka broker pod (not sure you can via Kafka CLI tools as it's an internal topic but it's more internal than the __consumer_offset one :-D).

Reply:
I will check that, thanks for the suggestion :)